File: apps/android/core/src/main/kotlin/ai/deepseek/dsh/companion/LiteStores.kt
package ai.deepseek.dsh.companion

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import java.io.File
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Durable Lite session identity: its stable id plus the event journal that
 * is the session — replaying the events through the fold reproduces the
 * whole state.
 */
class LiteSession(val id: String, events: List<JsonObject> = emptyList()) {
    private val journal = events.toMutableList()

    /** The recorded events, in journal order. */
    val events: List<JsonObject> get() = journal.toList()

    /** Append one event to the journal. */
    fun record(event: JsonObject) {
        journal.add(event)
    }

    /** Replay the journal through the Behavior-Spec fold. */
    val state: LiteDomainState
        get() {
            val fold = LiteFold()
            for (event in events) fold.apply(event)
            return fold.state
        }
}

/**
 * Persists Lite session journals — one append-only JSON-lines file per
 * session, one encoded event per line (plan chapter 11's journal shape).
 */
interface LiteSessionStoring {
    /**
     * Persist the session's complete journal, replacing any prior one.
     * @param session the session whose events become durable.
     */
    fun save(session: LiteSession)

    /**
     * Load one session's journal.
     * @param id the session identity.
     * @return the replayable session, or null when none is stored.
     */
    fun load(id: String): LiteSession?

    /**
     * Remove one session's journal.
     * @param id the session identity to delete.
     */
    fun delete(id: String)
}

/** File-backed session journals under a directory, `<id>.litejournal`. */
class LiteFileSessionStore(directory: File) : LiteSessionStoring {
    private val directory: File = directory.apply { mkdirs() }

    private fun file(id: String) = File(directory, "$id.litejournal")

    override fun save(session: LiteSession) {
        val text = session.events.joinToString("\n") { it.toString() } + "\n"
        writeAtomically(file(session.id), text.encodeToByteArray())
    }

    override fun load(id: String): LiteSession? {
        val file = file(id)
        if (!file.exists()) return null
        val session = LiteSession(id)
        for (line in file.readText().split('\n')) {
            if (line.isBlank()) continue
            val event = runCatching { Json.parseToJsonElement(line) as? JsonObject }.getOrNull()
                ?: error("corrupt journal line in ${file.name}")
            session.record(event)
        }
        return session
    }

    override fun delete(id: String) {
        file(id).delete()
    }
}

/**
 * Stores artifact content out-of-band (plan chapter 56): events carry only
 * references and status; bytes live in the resource channel.
 */
interface LiteArtifactStoring {
    /**
     * Write one artifact's bytes under its id.
     * @param id the artifact reference identity.
     * @param data the complete content bytes.
     */
    fun put(id: String, data: ByteArray)

    /**
     * Read one artifact's bytes.
     * @param id the artifact reference identity.
     * @return the stored bytes, or null when absent.
     */
    fun get(id: String): ByteArray?

    /**
     * Remove one artifact's bytes.
     * @param id the artifact reference identity.
     */
    fun remove(id: String)
}

/** File-backed artifact content under a directory, `<id>.artifact`. */
class LiteFileArtifactStore(directory: File) : LiteArtifactStoring {
    private val directory: File = directory.apply { mkdirs() }

    private fun file(id: String) = File(directory, "$id.artifact")

    override fun put(id: String, data: ByteArray) {
        writeAtomically(file(id), data)
    }

    override fun get(id: String): ByteArray? {
        val file = file(id)
        return if (file.exists()) file.readBytes() else null
    }

    override fun remove(id: String) {
        file(id).delete()
    }
}

/** Write bytes through a same-directory temp file and an atomic rename,
 * falling back to a plain replace where the platform refuses atomic moves. */
private fun writeAtomically(target: File, bytes: ByteArray) {
    val temp = File(target.parentFile, target.name + ".tmp")
    temp.writeBytes(bytes)
    try {
        Files.move(temp.toPath(), target.toPath(), StandardCopyOption.ATOMIC_MOVE)
    } catch (unsupported: AtomicMoveNotSupportedException) {
        // Nothing else can reach this catch: only the atomic flavor is
        // unsupported here, and the plain replace below is equivalent.
        Files.move(temp.toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING)
    }
}

File: apps/apple/Sources/LiteRuntime/LiteStores.swift
import Foundation

/// Durable Lite session identity.
public struct LiteSession: Equatable {
    public let id: String
    public private(set) var events: [LiteEvent]

    /// - Parameter id: stable session identity (also the storage key).
    public init(id: String, events: [LiteEvent] = []) {
        self.id = id
        self.events = events
    }

    /// Append one event to the journal.
    public mutating func record(_ event: LiteEvent) {
        events.append(event)
    }

    /// Replay the journal through the Behavior-Spec fold.
    public var state: LiteDomainState {
        var fold = LiteFold()
        for event in events { fold.apply(event) }
        return fold.state
    }
}

/// Persists Lite session journals — one append-only JSON-lines file per
/// session, one encoded event per line (plan chapter 11's journal shape).
public protocol LiteSessionStoring: Actor {
    /// Persist the session's complete journal, replacing any prior one.
    /// - Parameter session: the session whose events are durable.
    /// - Throws on storage failure.
    func save(_ session: LiteSession) async throws

    /// Load one session's journal.
    /// - Parameter id: the session identity.
    /// - Returns the replayable session, or nil when none is stored.
    /// - Throws on storage failure or a corrupt journal line.
    func load(id: String) async throws -> LiteSession?

    /// Remove one session's journal.
    /// - Parameter id: the session identity to delete.
    /// - Throws on storage failure.
    func delete(id: String) async throws
}

/// File-backed session journals under a directory, `<id>.litejournal`.
public actor LiteFileSessionStore: LiteSessionStoring {
    private let directory: URL
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    /// - Parameter directory: the directory holding one journal file per session.
    public init(directory: URL) throws {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        self.directory = directory
    }

    private func file(for id: String) -> URL {
        directory.appendingPathComponent("\(id).litejournal")
    }

    public func save(_ session: LiteSession) async throws {
        let lines = try session.events.map { event in
            String(data: try encoder.encode(event), encoding: .utf8)!
        }
        try lines.joined(separator: "\n").appending("\n").data(using: .utf8)!.write(to: file(for: session.id), options: .atomic)
    }

    public func load(id: String) async throws -> LiteSession? {
        let url = file(for: id)
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        var session = LiteSession(id: id)
        for line in text.split(separator: "\n") where !line.isEmpty {
            session.record(try decoder.decode(LiteEvent.self, from: Data(line.utf8)))
        }
        return session
    }

    public func delete(id: String) async throws {
        try? FileManager.default.removeItem(at: file(for: id))
    }
}

/// Stores artifact content out-of-band (plan chapter 56): events carry only
/// references and status; bytes live in the resource channel.
public protocol LiteArtifactStoring: Actor {
    /// Write one artifact's bytes under its id.
    /// - Parameters:
    ///   - id: the artifact reference identity.
    ///   - data: the complete content bytes.
    /// - Throws on storage failure.
    func put(id: String, data: Data) async throws

    /// Read one artifact's bytes.
    /// - Parameter id: the artifact reference identity.
    /// - Returns the stored bytes, or nil when absent.
    /// - Throws on storage failure.
    func get(id: String) async throws -> Data?

    /// Remove one artifact's bytes.
    /// - Parameter id: the artifact reference identity.
    /// - Throws on storage failure.
    func remove(id: String) async throws
}

/// File-backed artifact content under a directory, `<id>.artifact`.
public actor LiteFileArtifactStore: LiteArtifactStoring {
    private let directory: URL

    /// - Parameter directory: the directory holding one content file per artifact.
    public init(directory: URL) throws {
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        self.directory = directory
    }

    private func file(for id: String) -> URL {
        directory.appendingPathComponent("\(id).artifact")
    }

    public func put(id: String, data: Data) async throws {
        try data.write(to: file(for: id), options: .atomic)
    }

    public func get(id: String) async throws -> Data? {
        try? Data(contentsOf: file(for: id))
    }

    public func remove(id: String) async throws {
        try? FileManager.default.removeItem(at: file(for: id))
    }
}
